android/src/main/java/com/auth0/react/A0Auth0Module.java

```diff
@@ -52,16 +54,15 @@ public String getName() {
     @ReactMethod
     public void showUrl(String url, boolean closeOnLoad, Callback callback) {
         final Activity activity = getCurrentActivity();
-
+        final Uri parsedUrl = Uri.parse(url);
         this.callback = callback;
+
         if (activity != null) {
-            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
-            CustomTabsIntent customTabsIntent = builder.build();
-            customTabsIntent.launchUrl(activity, Uri.parse(url));
+            AuthenticationActivity.authenticateUsingBrowser(activity, parsedUrl);
         } else {
             final Intent intent = new Intent(Intent.ACTION_VIEW);
             intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
-            intent.setData(Uri.parse(url));
+            intent.setData(parsedUrl);
             getReactApplicationContext().startActivity(intent);
```

Comment:
how can we ensure an "activity context" is always present? would this branch still work or have the same bug?

Reply:
Isn't that betond the scope of this PR?

Reply:
Is not out of scope if the other branch still has this bug. Was that verified?

Reply:
But that branch just opens whatever browser the device has installed. That's another process.
Should I check that anyway?

Reply:
Let me rephrase my original question. If an "Activity context" is not present (activity == null), and so the "else" branch is executed. Would the "authentication restart" bug still occur? and how was that verified?

I'm not familiar with the reasons why the activity is being null-checked. It was [added 4y ago](https://github.com/auth0/react-native-auth0/commit/b841aeb9e1338da6bfee3a6f54ec475d8e3d7c78). I suppose devs were calling the `showUrl()` method from places other than an Activity and this React `getCurrentActivity()` method failed to find one. If having an activity is a requirement for the code, we should document that in the usage/instructions.

Other than using an Activity or an Application as the context, seems like both will end up executing an ACTION_VIEW intent passing the URL as data. Because [CustomTabsIntent.launchUrl(context, url)](https://developers.google.com/web/android/custom-tabs/implementation-guide#what_happens_if_the_user_doesn%E2%80%99t_have_a_browser_that_supports_custom_tabs_installed) will default to that same behavior according to the docs. Does it make sense to maintain 2 separate if/else branches when it can be reduced to just delegating the authentication to the new `AuthenticationActivity`?

Reply:
Indeed, the code needs an Activity to work. I'll add a couple of lines to the Readme.
The `@react-native-community/google-signin` package [straight up produces an error](https://github.com/react-native-google-signin/google-signin/blob/master/android/src/main/java/co/apptailor/googlesignin/RNGoogleSigninModule.java#L189) if the Activity is `null`. And the Activity will be `null` when [there's no Activity attached](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/ReactContext.java#L422) to the current `ReactContext`. And Given that CustomTabs will fallback to a regular browser, it makes no sense to keep the else branch. I'll remove it.

Reply:
> And given that CustomTabs will fallback to a regular browser, it makes no sense to keep the else branch. I'll remove it

👍 

> the code needs an Activity to work.

Let's make this clear in the usage section of the README. Also, if we fail like the linked package above does, let's use a custom error code as we do in other places (e.g. "a0.internal.invalid_android_context") so the developers can match against this code and handle the error better.

Reply:
Done in https://github.com/auth0/react-native-auth0/pull/350/commits/68b09f3d0360732eb71d7f0bf72fdf4ecff21b3d.